lib/core/src/display/symbol/mesh.rs

```diff
@@ -86,163 +146,106 @@ impl<OnDirty:Callback0+Clone> Mesh<OnDirty> {
-
-
-
-        mat4 model_view_projection_matrix = view_projection_matrix * model_matrix;
-
-
-        local       = vec3((uv - 0.5) * bbox, 0.0);
-        gl_Position = model_view_projection_matrix * vec4(local,1.0);
-
-
-
-
-        _position = view_projection_matrix * vec4(local,1.0);
-
-
-//        mat4 model_view_matrix  = view_matrix * model_matrix;
-//        vec4 eyeT               = model_view_matrix * vec4(local,1.0);
-//        gl_Position             = projection_matrix * eyeT;
-//        world                   = gl_Position.xyz;
-//        eye                     = eyeT.xyz;
-//        eye.z                   = -eye.z;
-
-    }
-"#,
-            )
-                .unwrap();
-            let frag_shader = webgl::compile_shader(
-                &self.context,
-                webgl::Context::FRAGMENT_SHADER,
-                r#"#version 300 es
-
-    precision highp float;
-    precision highp int;
-
-    out vec4 out_color;
-    in mat4 model_matrix;
-    in vec4 _position;
-    in vec2 uv;
-
-    in vec3 local;
-
-    uniform mediump mat4 view_matrix;
-    uniform mediump mat4 projection_matrix;
-    uniform mediump mat4 view_projection_matrix;
-
-    void main() {
-        out_color = vec4(uv.x, uv.y, 0.0, 1.0);
-        out_color = vec4(local, 1.0);
-        out_color = _position;
-        out_color = vec4(1.0,1.0,1.0,1.0);
+    /// Creates a new VertexArrayObject, discovers all variable bindings from material to geometry,
+    /// and initializes the VAO with the bindings.
+    fn init_vao(&mut self) {
+        self.vao = Some(VAO::new(&self.context));
+        self.with_program(|program|{
+            let var_bindings = self.discover_variable_bindings();
+            for (variable,opt_scope_type) in &var_bindings {
+                if let Some(scope_type) = opt_scope_type {
+                    let opt_scope = self.geometry.var_scope(*scope_type);
+                    match opt_scope {
+                        None => self.logger.error("Internal error. Invalid var scope."),
+                        Some(scope) => {
+                            let vtx_name = shader::builder::mk_vertex_name(&variable);
+                            let location = self.context.get_attrib_location(program, &vtx_name);
+                            if location < 0 {
+                                self.logger.error(|| format!("Attribute '{}' not found.",vtx_name));
+                            } else {
+                                let location     = location as u32;
+                                let buffer       = &scope.buffer(&variable).unwrap();
+                                let is_instanced = scope_type == &geometry::ScopeType::Instance;
+                                buffer.bind(webgl::Context::ARRAY_BUFFER);
+                                buffer.vertex_attrib_pointer(location, is_instanced);
+                            }
+                        }
+                    }
+                }
+            }
+        });
```

Comment:
This function have a way too long indent. It should be split to the subfunctions.

Reply:
I was thinking about. Moreover I already did the refactor, but it was ugly as hell, as the nested thing uses almost every variable from outer scope. Can I ask you for help here?